(Breaking) API changes

Tederean opened the pull request:
I took a few hours today to think about the source code. I have made some changes, whereby I have only adapted everything in the winrt backend to the new status. I still have to update Linux and macOS, but I wanted to get some feedback first.

1. Unavailable values because of Backend, private DeviceID content
Not all backends can handle serial numbers. WebHID and WinRT cannot, with all others it is optional if it is a USB device that actually has a serial number. If only the serial number was optional, I would like the SerielNumberExt trait, but with WebHID, UsagePage and UsageId are also not available.
Then it bothers me that I cannot call up the content of the DeviceID from external source code. I had a look at how nusb crate does this, they offer everything directly, but everything that is not available in the current configuration is simply hidden. I am of the opinion that there is too much wrapping going on in async-hid here.

2. Filter devices while enumerating
WebHID requires us to specify which devices we would like to have, ‘simply all’ is not possible. So we need a new additional method.

3. Split Device Reader & Writer, use &mut self
- Eliminates the error of writing with a device that can only read.
- Makes many synchronisation problems obsolete.

Please take a look at the source code, especially lib.rs. Many files are simply affected by rustfmt. I think this API solves a lot of problems or causes them at compile time. It is much clearer which information is available when.


I welcome constructive criticism and suggestions.

sidit77 commented:
I agree that the `SerialNumberExt` should go. However, I would just always set the serial number to `None` on platforms that don't support serial numbers and document that fact. UsagePage and UsageId are available in WebHID inside of the [collections](https://developer.mozilla.org/en-US/docs/Web/API/HIDDevice/collections) property, so this shouldn't be an issue.

I really want to keep the properties of `DeviceInfo` exposed. Let's assume you have some software that can manage 30 different devices. If we remove all the pid/vid/usage/page info it would be impossible to tell which "driver" you should use for a found device without enumerating 30 times with different filters. I don't think we need filters at all tbh. In WebHID the filters are only used to customize the permission dialog shown to the user, so I wouldn't phrase it as a general filter but rather as something that customizes the web permission dialog and is ignored on all other platforms. It's not even required for enumeration as long as you've given the website access before:
![image](https://github.com/user-attachments/assets/2c425caf-a8df-4e92-b806-f64c52758a57)

 I would also like to keep the opaque `DeviceId` around, as you can compare it to other `DeviceId`s for example or print it for debug purposes. The type is intentionally nested to hide the platform-specific type from the [public api](https://docs.rs/async-hid/latest/async_hid/struct.DeviceId.html).

For the read/write api, I would prefer to keep them together by default and have a `fn split(self)->(Reader, Writer)` function to panics if the device wasn't opened in ReadWrite mode as it isn't relevant for all other modes. But I also have to think about this a little bit more.

Tederean commented:
> I agree that the SerialNumberExt should go. However, I would just always set the serial number to None on platforms that don't support serial numbers and document that fact.

My problem with None is that it is ambiguous in this case. None because the device has no serial number or None because this feature isn't supported. Of course you can document this, but the idea behind my approach was that overlooking this documentation information would lead to a compilation error and not to surprises at runtime because I hadn't thought of it.



> UsagePage and UsageId are available in WebHID inside of the [collections](https://developer.mozilla.org/en-US/docs/Web/API/HIDDevice/collections) property, so this shouldn't be an issue.

> [Filters] It's not even required for enumeration as long as you've given the website access before

I haven't seen that yet, I was previously of the opinion that this is only possible via a filter. I'm going to try both out. I would be happy if these adjustments were not necessary, I personally don't need a separate filter either.  As we still have a few more construction sites, I would suggest we leave any WebHID adjustments aside for now.



> I really want to keep the properties of DeviceInfo exposed.

Me too. I think there was a slight misunderstanding here. I want to keep all information from DeviceInfo open, but via methods and not fields (see below in lib.rs).


> I would also like to keep the opaque DeviceId around, as you can compare it to other DeviceIds for example or print it for debug purposes. The type is intentionally nested to hide the platform-specific type from the [public api](https://docs.rs/async-hid/latest/async_hid/struct.DeviceId.html).

I get nothing out of it except that I can compare whether it is the same device. async-HID is still missing features. That's not bad, unless I lock everything. I really don't think it's a good idea to withhold the handles/IDs/paths from the developer. Before I developed my own HID device, I worked with a device that queried custom HID attributes under Windows via win32, for which you need a Windows handle (HString). This was because of a custom driver from manufacturer providing additional metadata. It really sucks when the HString is right under your nose, but the developer of the library patronises you. So I have to build a separate implementation that lists all devices, filtering again just to get information that is already there.

DeviceId(u64) on the Mac, for example, is also not very descriptive in the source code, but registry_entry_id() -> u64.


> For the read/write api, I would prefer to keep them together by default and have a fn split(self)->(Reader, Writer) function to panics if the device wasn't opened in ReadWrite mode as it isn't relevant for all other modes. But I also have to think about this a little bit more.

We could do it that way, that was my idea at first too. And then I realised how many avoidable panic situations this creates if you just let it go. For the Struct what Reader and Writer holds you generate following panic abilities:
- write can panic when there is no Writer
- read can panic if there is no Reader
- split must then return (Option<Reader>, Option<Writer>) which forces me as a library user to unwarp 1 to 2 times.

Or I simly do it like this as a library user. Fits perfectly into the rust type system and is KISS:

```
pub struct Device {
  pub reader: DeviceReader,
  pub writer: DeviceWriter,
}

let (reader, writer) = device_info.open()?;

let device = Device {
  reader,
  writer,
};
```

I have now removed cfg's of UsagePage & UsageID and also removed the filtering.
Please let me know what you think about my objections.
